### ISSUE TITLE
idea for coastline subsetting

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1498]
See attached (a document that is in the `sandbox/coastlineSubset` directory of the repo's "develop" branch. Any thoughts from co-developers and users?

[coastline_clip.pdf](https://github.com/dankelley/oce/files/2894780/coastline_clip.pdf)

**Comment by dankelley** [comment]
Hm, or I could just google for it...

“Sutherland–Hodgman Algorithm.” In Wikipedia, March 4, 2018. https://en.wikipedia.org/w/index.php?title=Sutherland%E2%80%93Hodgman_algorithm&oldid=828823016.

**Comment by dankelley** [comment]
I'm a bit too lazy to read those papers I cited, but the idea seems to be what I had in mind, so I coded that as `sandbox/coastlineSubset/02.R` for an interactive test that suggests my method now just needs to be extended for all 4 sides of a box.

It's quite slow in R, but if I really become convinced that my logic is okay, it could be done in C for speed.  I've done this kind of thing manually for years, and I really ought to have coded it instead.

It could go into oce, actually -- I think if people are using `mapPlot` they will appreciate the speed, although probably `plot,coastline-method` is so fast that people wouldn't care.

**Comment by dankelley** [comment]
Don't tell the orange-haired one, but the USA and Venezuela both disappear from my map, with my still-broken code:
<img width="793" alt="screen shot 2019-02-24 at 11 46 06 am" src="https://user-images.githubusercontent.com/99469/53301499-f367ee00-3829-11e9-810c-7c1357fa5a5b.png">

**Comment by dankelley** [comment]
I think `sandbox/coastlineSubset/04.R` works now. It prints quite a lot of debugging stuff and plots a lot of panels. I invite others to play with it, to see if they can break it, and to send me a screenshot if they find a faulty result.

To test it, just source that file, then activate the plot window and click once at the lower-left (southwest) corner of the focus box, then a second time at the upper-right (northeast) corner. It will then draw panels for the 4 steps of "cutting" (first at right, then below, then to left, and finally at top). And then the very final plot, in the bottom-right of the 6 panels, will be the results.

Note that the countries (or islands or whatever) are colourized, to help a bit in seeing them. Howeve, I am not doing a 4-colour map calculation, but simply cycling through colours, and the regions are in a funny order, so e.g. the US and Canada will both come out as blue, if you have enough countries in central and south America included.

So, @clayton33 and @richardsc (and anyone else) -- try to break it!

This is actually not crazy-slow, and since this would be done rarely, I could put it into oce as an R code (so people can fix errors if they find them).  I would have used this perhaps twice a year since I learned R, actually, so I do think it ought to go into oce, and might do that on the weekend, unless someone objects.

**Comment by clayton33** [comment]
Not sure if this is classified as 'breaking' but the final picture is kinda funny looking.

![coastlinesubset01](https://user-images.githubusercontent.com/3463161/53358667-53809200-3907-11e9-9552-ec18d7e608ac.PNG)

**Comment by dankelley** [comment]
Definitely breaking!  Looks like there is another special case I need to consider. Thanks!

**Comment by richardsc** [comment]
Here's my first attempt. Not as bad as Chantelle's, but there is a little weirdness in the upper right hand corner:
![image](https://user-images.githubusercontent.com/233584/53360335-a9573900-390b-11e9-96db-e8c5c706eb30.png)
```
RIGHT: trim to west of E (starting with 290 segments)
BOTTOM: trim to north of S (starting with 134 segments)
LEFT: trim to east of W (starting with 113 segments)
TOP: trim to south of N (starting with 46 segments)
range(LON)=0.772335646171484 3.79711225751171 (W=-20.757271463281, E=12.106512366052)
range(LON)=-5.47056494792901 2.17710778159378 (W=-20.757271463281, E=12.106512366052)
range(LON)=-8.60288021486862 -2.56218950032624 (W=-20.757271463281, E=12.106512366052)
range(LON)=8.48881554529089 12.106512366052 (W=-20.757271463281, E=12.106512366052)
adding initial NE vertex at N=27.3560095379839 E=12.106512366052
range(LON)=-8.62267579949546 11.9995056494716 (W=-20.757271463281, E=12.106512366052)
range(LON)=9.05709629547272 12.106512366052 (W=-20.757271463281, E=12.106512366052)
range(LON)=-3.24437008301126 1.06012169760493 (W=-20.757271463281, E=12.106512366052)
range(LON)=-15.1303112451682 -7.83210038901919 (W=-20.757271463281, E=12.106512366052)
range(LON)=-16.8415246240813 -13.8449633447724 (W=-20.757271463281, E=12.106512366052)
range(LON)=-16.6774519515546 -13.7004760400843 (W=-20.757271463281, E=12.106512366052)
range(LON)=9.30561323409626 11.2850789730365 (W=-20.757271463281, E=12.106512366052)
range(LON)=-11.4387794661821 -7.53971513511176 (W=-20.757271463281, E=12.106512366052)
range(LON)=9.31941084151816 11.9995056494716 (W=-20.757271463281, E=12.106512366052)
range(LON)=-17.0204284326757 -8.79488399904908 (W=-20.757271463281, E=12.106512366052)
range(LON)=-12.1707502913803 4.2702099951438 (W=-20.757271463281, E=12.106512366052)
range(LON)=-17.0634232243426 -4.92333736817423 (W=-20.757271463281, E=12.106512366052)
range(LON)=0.295646396495101 12.106512366052 (W=-20.757271463281, E=12.106512366052)
range(LON)=2.69170169435625 12.106512366052 (W=-20.757271463281, E=12.106512366052)
range(LON)=-17.0634232243426 -8.68447570155024 (W=-20.757271463281, E=12.106512366052)
range(LON)=-17.6250426904907 -11.4678991357785 (W=-20.757271463281, E=12.106512366052)
range(LON)=-13.2465502588325 -10.2300935530913 (W=-20.757271463281, E=12.106512366052)
range(LON)=-0.0497847151599444 1.86524051271232 (W=-20.757271463281, E=12.106512366052)
Sourced file /home/richardsc/src/oce/sandbox/coastlineSubset/04.R
Warning messages:
1: In plot,coastline-method() : should trim poles
 
2: In plot,coastline-method() : should trim poles
 
3: In plot,coastline-method() : should trim poles
 
4: In plot,coastline-method() : should trim poles
 
5: In plot,coastline-method() : should trim poles
 ```

**Comment by richardsc** [comment]
Another break. This one looks a lot like Chantelle's. The CAA seems to be an issue with this (maybe because of all the islands?
![image](https://user-images.githubusercontent.com/233584/53360495-08b54900-390c-11e9-8578-4825dd79ae35.png)
RIGHT: trim to west of E (starting with 290 segments)
BOTTOM: trim to north of S (starting with 52 segments)
LEFT: trim to east of W (starting with 25 segments)
TOP: trim to south of N (starting with 17 segments)
range(LON)=-86.065634306799 -84.1952624482337 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-99.7974012038005 -95.6476812038005 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-105.444714408101 -82.6928640647163 (W=-105.444714408101, E=-82.6928640647163)
adding initial NW vertex at N=80.4084216471672 W=-105.444714408101
range(LON)=-105.444714408101 -100.98078 (W=-105.444714408101, E=-82.6928640647163)
adding initial NW vertex at N=80.4084216471672 W=-105.444714408101
range(LON)=-105.444714408101 -104.5 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-90.205160285182 -82.6928640647163 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-102.5 -96.54 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-96.0337450833824 -90.5097928535426 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-96.8209321764845 -93.6127559069405 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-102.56552 -97.704415 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-97.1213789538295 -82.6928640647163 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-96.4363044909361 -93.7206562975659 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-98.6319844225855 -95.5592779202946 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-105.444714408101 -99.6709390938136 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-96.70972 -85.81435 (W=-105.444714408101, E=-82.6928640647163)
range(LON)=-89.6161 -82.6928640647163 (W=-105.444714408101, E=-82.6928640647163)

**Comment by richardsc** [comment]
And .... another:
![image](https://user-images.githubusercontent.com/233584/53360620-63e73b80-390c-11e9-9df5-ae680a278fa5.png)
RIGHT: trim to west of E (starting with 290 segments)
BOTTOM: trim to north of S (starting with 218 segments)
LEFT: trim to east of W (starting with 110 segments)
TOP: trim to south of N (starting with 10 segments)
range(LON)=60.5284298033116 62.6661797957952 (W=48.7622712526158, E=62.6661797957952)
range(LON)=48.7622712526158 50.3928210793127 (W=48.7622712526158, E=62.6661797957952)
adding initial NW vertex at N=50.0927575847767 W=48.7622712526158
range(LON)=48.7622712526158 61.2108170917257 (W=48.7622712526158, E=62.6661797957952)
range(LON)=48.7622712526158 62.6661797957952 (W=48.7622712526158, E=62.6661797957952)
adding final NW vertex at N=50.0927575847767 W=48.7622712526158
range(LON)=48.7622712526158 49.1011600000001 (W=48.7622712526158, E=62.6661797957952)
adding initial NW vertex at N=50.0927575847767 W=48.7622712526158
range(LON)=52.5024597511961 62.6661797957952 (W=48.7622712526158, E=62.6661797957952)
adding initial NE vertex at N=50.0927575847767 E=62.6661797957952
range(LON)=55.9289172707411 62.6661797957952 (W=48.7622712526158, E=62.6661797957952)
adding initial NE vertex at N=50.0927575847767 E=62.6661797957952

**Comment by clayton33** [comment]
I found another.
![coastlinesubset02](https://user-images.githubusercontent.com/3463161/53361269-c1c85300-390d-11e9-9b24-c5d0366da3cd.PNG)

RIGHT: trim to west of E (starting with 290 segments)
BOTTOM: trim to north of S (starting with 275 segments)
LEFT: trim to east of W (starting with 52 segments)
TOP: trim to south of N (starting with 5 segments)
range(LON)=106.845062487955 149.541458497028 (W=106.845062487955, E=149.541458497028)
range(LON)=139.86312 143.60385 (W=106.845062487955, E=149.541458497028)
range(LON)=146.11919 149.541458497028 (W=106.845062487955, E=149.541458497028)
range(LON)=136.97439 145.086285 (W=106.845062487955, E=149.541458497028)

**Comment by dankelley** [comment]
I'm going to look into doing this with `sp`, `rgeos`, etc ... I think there are many methods possible, so I'll try to find one that is simple, and that I find pleasing.  (There are a lot of packages about things relating to spatial analysis.)

**Comment by richardsc** [comment]
Two things:

1. It might be worth an email to the R-SIG-GEO mailing list to see what ideas those folks have (the experts are all on there).

2. Will the intended functionality of this also apply to `mapPlot()` (and not just `plot,coastline-method`?

**Comment by dankelley** [comment]
1. My browser has about 6 tabs open to solutions. I think there are tons of solutions out there.

2. This is (so far) just intended for letting people subset large coastline files for plot speed. It has no relationship to any plotting functions. In my tests, coastline plotting is quite fast even for `data(coastlineWorldFine,package="ocedata")` but `mapPlot` is annoyingly slow.  We did this kind of thing for SLEIWEX, with coastlines from, well, I don't know where, but possibly hand-digitized charts.

**Comment by dankelley** [comment]
The below seems sensible:
```R
## This is as at
## https://www.rdocumentation.org/packages/raster/versions/2.8-19/topics/intersect
## with some small extras by DK.
library(raster)
e1 <- extent(-10, 10, -20, 20)
e2 <- extent(0, 20, -40, 5)
##intersect(e1, e2)
par(mfrow=c(1, 2), mar=rep(1, 4))

#SpatialPolygons
if (require(rgdal) & require(rgeos)) {
	p <- shapefile(system.file("external/lux.shp", package="raster"))
	b <- as(extent(6, 6.4, 49.75, 50), 'SpatialPolygons')
	projection(b) <- projection(p)
	i <- intersect(p, b)
	plot(p)
        box()
	plot(b, add=TRUE, col='red') # excluded part
	plot(i, add=TRUE, col='blue', lwd=2) # included part
}
## DK: let'sreplot to check -- yup, looks great.
plot(i, col=1:12)
box()
## DK: next, how do we get those numbers out of 'i'?
```

**Comment by dankelley** [comment]
Second test request to @clayton33 and @richardsc : please try `sandbox/coastlineSubset/05.R` and report any errors. It works the same: click the mouse twice to define a box.  And, again, the colors may repeat on neighbors.

The code is pretty funky in spots. I could not figure out the proper way to extract the coords from the object (nothing I tried based on things I read online seemed to work) so I just inspected the object to find what is where ... risky but ok for a test.

As before, if you can break this, provide a screenshot so I can check it.

**Comment by richardsc** [comment]
Can you update so that after the subset a new plot is made with *only* the subsetted region? I think I've broken it, but I find it hard to tell without seeing a plot of the new region (instead of just the "overlay"):

Notice how Quebec doesn't get colored in this one:
![image](https://user-images.githubusercontent.com/233584/53425071-c139c600-39ba-11e9-9b30-ceecb1d0ee9f.png)

**Comment by richardsc** [comment]
Here's another, where only part of Ellesmere seem to have gotten snipped properly:
![image](https://user-images.githubusercontent.com/233584/53425193-f6deaf00-39ba-11e9-9878-4aa773ee38ab.png)

**Comment by dankelley** [comment]
It's updated (commit afa30846620ff0a60510315d528dcc93fda7d592). Oh, jeeze, I was thinking that it would all work 100% if we used these well-tested packages.  But I get a badness as you can see below.  (Has Québec left the world?).  I **really** don't want to try to get this working with the half-dozen or so methods that seem to exist... frustrating.

![screen shot 2019-02-26 at 12 45 13 pm](https://user-images.githubusercontent.com/99469/53430176-64430d80-39c4-11e9-9963-68462a341e0e.png)

**Comment by dankelley** [comment]
Oh, this is weird. If you do 'git pull' you'll get a version in which the first line is
```
REAL <- !TRUE # to avoid having to click whilst writing code
```
(and some other small changes).  I had to look further within the polygons to find what I wanted. In this example, I now get Québec, and Maine, too!  But ... I'm missing part of Nunavut (look at the top edge of the box ... some land didn't get coloured in). I don't know what the problem is.  I have more than just 2 points in there. Maybe it's because there is only a single point that's north of the box.

![screen shot 2019-02-26 at 1 39 00 pm](https://user-images.githubusercontent.com/99469/53433938-e420a600-39cb-11e9-91c5-cf4604fd79c0.png)

**Comment by dankelley** [comment]
Oh, just shoot me now. I was using a "for i in length(blah)" instead of "seq_along(blah)". Now, I think it works perfectly!  Maybe. And maybe not.  (Insert Maxwell Smart quote here, I think.)

**Comment by dankelley** [comment]
I think `sandbox/coastlineSubset/05.R` is working nicely now. You need to change the assigned value on line 1 to re-actualize the mouse action. Also, now, it's using the fine-res dataset. Oh, and it spits out `coastline.rda` at the end ... you can tell I am now doing steps to make this into a function, the docs of which would say how to use the mouse to do this graphical work, etc.

The snapshot shows what it looks like, and the R session that's visible shows the time taken to plot this `coastline.rda` dataset is *way* less than to plot the whole dataset. (This may vary with the plot device and memory on the machine, but even just the number of points says it ought to be 37 times faster.)

<img width="533" alt="screen shot 2019-02-26 at 4 39 30 pm" src="https://user-images.githubusercontent.com/99469/53444821-8436f900-39e5-11e9-8ec2-f95405f2f293.png">

**Comment by richardsc** [comment]
I have done this a dozen times or so over the last day, and have not seen a breaking case.

**Comment by dankelley** [comment]
Good. I plan to add a `generic` method for `coastline` so a person can do e.g.
```R
cl <- subset(coastlineWorldFine, -70 < longitude & longitude < -30 & 30 < latitude & latitude < 50)
```
because then users will get a documented way to do this.  I'll close this issue when I'm done, likely on the weekend.

The other reason I was so keen to get this working was because I think it will lead the way to solve other coastline problems. I am not going to raise issues on them, though ... I will likely just play a bit until I see if there's any promise to my ideas.

**Comment by dankelley** [comment]
This method has been incorporated in the "develop" branch, so I am closing this issue. To see how the method works, try `examples("subset,coastline-method")` or just copy the example code from `?"subset,coastline-method"` into an R console.

Thanks to @clayton33 and @richardsc for the extensive help with testing!

**Comment by dankelley** [comment]
TL;DR I plan to make `subset,coastline-method` require 2nd arg to be a string.

I've reopened this for a couple of hours, because I plan to insist that the subset be a string. Why? Because I think it will be too confusing for users, otherwise. For example, in our domain part of the expression might be `longitude< -50`. That seems straightforward. But a careful analyst might write this as `longitude<(-50)`, just to be sure that it would not turn into assignment. Now, then, you ask, when you call a generic `subset()` function, and part of the stated `subset` arg contains `<-`, does that do an assignment?  Well, yes, it does.

The other thing is that it is quite hard to work with it programmically, e.g. consider the following (which is what I'm actually trying to do, to be honest ... it's in my effort to strengthen `coastlineCut()`)
```R
cut <- 100
subset(coastlineWorld, -180<longitude&longitude<cut&-90<latitude&latitude<90)
```
Well, I'm having a hard time figuring out how to make `subset()` "reach back" into the calling program to determine the value of this `cut` thing.  We cannot try to "reach back" for the whole expression, because e.g. `longitude` might be defined there, and we definitely do *not* want to use the value from there. In other words, we want to pass "longitude" and "latitude" through as symbols, without evaluation, but then we *do* want to evaluate things to the left or right of the `<` symbols.

If it's hard for me to code, that means it's kind of hard for the user, perhaps.  And, anyway, I don't think it is really too much to ask users to put quotes around the expression. If it's just a simple expression with constant limits, that's trivial. And if not, e.g. as in the `coastlineCut()` case, it's still pretty
easy:
```R
cut <- 100
sub <- sprintf("-180<longitude&longitude<%f&-90<latitude&latitude<90", cut)
subset(coastlineWorld, sub)
```

**Comment by dankelley** [comment]
Actually, laundry now sorted (hm, how did I end up buying 12 identical pairs of socks when I thought I was ordering 3?), I don't like calling this `subset` at all. It is not really a clean generic, in my view. For example, the code below (exercise for the reader: can you guess what it plots?) shows that readers might expect to be able to give complicated statements for the second argument to `subset`. But th present `subset,coastline-method` does not offer that.  A simpler name, and something that will likely be easier for the user and a **lot** easier for the programmer, would be to have a `coastlineTrim(coastline,E,W,S,N)` function. So that's what I'll do, after a bit of unch-lae.

```R
x <- 1:20
y <- 1 + 2 * x
l <- data.frame(x=x,y=y)
plot(l)
lines(l$x, 30-l$x)
hilite <- subset(l, y < 30-x)
points(hilite, col=2)
```

**Comment by dankelley** [comment]
Author: dankelley <kelley.dan@gmail.com>
Date:   Sun Mar 3 13:49:59 2019 -0400

    coastlineTrim replaces subset,coastline-method()
    
    The reasoning is touched upon at
    https://github.com/dankelley/oce/issues/1498.
